apps/src/p5lab/spritelab/commands/criterionCommands.js

```diff
@@ -583,9 +583,92 @@ export const commands = {
+    const newEffects = effects.filter(
+      effect => effect.renderFrame === this.currentFrame()
+    );
+    return newEffects.length >= min;
+  },
+
+  // Returns true if a minimum number of variable bubbles are currently being drawn.
+  variableBubblesCreated(min = 1) {
+    return this.getVariableBubbles().length >= min;
+  },
+
+  // Returns true if a minimum number of watched variables have a valid value.
+  // If unset, a variable's value is an empty string.
+  variableValueSet(min = 1) {
+    const studentVars = this.getVariableBubbles();
+    const filteredVars = studentVars.filter(
+      studentVar => this.getVariableValue(studentVar.name) !== ''
+    );
+    return filteredVars.length >= min;
+  },
+
+  // Returns true if a watched variable's value changed this frame.
+  // Optionally checks if the variable values have increased or decreased.
+  // Since this command requires a variable bubble, it should be used in conjunction
+  // with variableBubblesCreated.
+  variableValueChanged(changeType = '') {
+    const previousStudentVars = this.previous.variableBubbles;
+    if (previousStudentVars) {
+      const studentVars = this.getVariableBubbles();
+      for (const studentVar of studentVars) {
+        // Find a variable with the same name in the previous student variables
+        const matchingPreviousVar = previousStudentVars.find(
+          previousVar => previousVar.name === studentVar.name
+        );
+
+        // If the variable existed, compare its value with the value of the current student variable
+        if (matchingPreviousVar) {
+          const previousValue = matchingPreviousVar.value;
+          const currentValue = this.getVariableValue(studentVar.name);
+
+          // Compare the values
+          if (previousValue !== currentValue) {
+            switch (changeType) {
+              case '':
+                // If we're not explicitly looking for an increase or decrease,
+                // any change to the value satisfies this criterion.
+                return true;
+              case 'decrease':
+                if (
+                  typeof previousValue === 'number' &&
+                  typeof currentValue === 'number' &&
+                  currentValue < previousValue
+                ) {
+                  return true;
+                }
+                break;
+              case 'increase':
+                if (
+                  typeof previousValue === 'number' &&
+                  typeof currentValue === 'number' &&
+                  currentValue > previousValue
+                ) {
+                  return true;
+                }
+                break;
+              default:
+                console.error(
+                  `Invalid argument for variableValueChanged: ${changeType}`
+                );
+                break;
+            }
+          }
+        }
+      }
+    }
+    // If we reach this point, none of the watched variables changed
```

Comment:
comment nit: they also could have changed, but not the way we are testing for, right?

Reply:
That’s fair. I can clarify.